### PR TITLE
Feature/kitosudv 2219 task 2317

### PR DESCRIPTION
**Pull request by mrjsawdk**
## KITOS Pull request template
The following procedure dictates the steps needed before a Pull request can be merged into master.

- [x] **Implement**: 
      *All requirements are implemented and unit tests are green*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*
      
- [x] **Request review**: 
      *Tag whomever you wish to review your code*
      
- [x] **Review completed**: 
      *Reviewer ticks this box when review comments have been submitted*
      
- [x] **Changes**: 
      *PR owner and reviewer agrees on which changes must be made and the changes are committed.*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*

### CODE REVIEW

File: Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs [hunk 1]
```diff
@@ -529,6 +531,134 @@ private CreateNewContractRequestDTO CreateNewSimpleRequest(Guid organizationUuid
             };
         }
 
+        [Theory]
+        [InlineData(true, true, true)]
+        [InlineData(true, false, true)]
+        [InlineData(false, true, true)]
+        public async Task Can_POST_With_GeneralData(bool withContractType, bool withContractTemplate, bool withAgreementElements)
```
Comment (JacobMalling) [inline Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs:538]: Ingen test uden AgreementElements?

File: Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs [hunk 2]
```diff
@@ -546,7 +676,7 @@ private string CreateEmail()
             return $"{nameof(ItContractsApiV2Test)}{A<string>()}@test.dk";
         }
 
-        private static void AssertCrossReference<T>(T expected, IdentityNamePairResponseDTO actual) where T : IHasNameExternal, IHasUuidExternal
+        private static void AssertCrossReference<TExpected, TActual>(TExpected expected, TActual actual) where TExpected : IHasNameExternal, IHasUuidExternal where TActual : IHasNameExternal, IHasUuidExternal
```
Comment (JacobMalling) [inline Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs:679]: Nogen grund til to generiske typer når de indeholder begge dele? Udover at øge læsbarheden

File: Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs [hunk 3]
```diff
@@ -235,6 +249,284 @@ public void Cannot_Create_With_Parent_If_Different_Organizations()
...
+        [InlineData(true, true, false, true, true)]
+        [InlineData(true, false, true, true, true)]
+        [InlineData(false, true, true, true, true)]
+        [InlineData(false, false, false, false, false)]
+        public void Can_Create_With_GeneralData(bool withContractType, bool withContractTemplate, bool withAgreementElements, bool withValidFrom, bool withValidTo)
+        {
+            // Arrange
+            var (organizationUuid, itContractModificationParameters, createdContract, transaction) = SetupCreateScenarioPrerequisites();
+
+            var contractId = A<string>();
```
Comment (JacobMalling) [inline Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs:265]: Der kan ikke laves en hjælpemetode til dette? Vil gøre det noget lettere når vi skal lave de to "store" tests

File: Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs [hunk 4]
```diff
@@ -235,6 +249,284 @@ public void Cannot_Create_With_Parent_If_Different_Organizations()
...
+
+
+            // Act
+            var result = _sut.Create(organizationUuid, itContractModificationParameters);
+
+            // Assert
+            Assert.True(result.Ok);
+            AssertTransactionCommitted(transaction);
+            var contract = result.Value;
+            Assert.Equal(contractId, contract.ItContractId);
```
Comment (JacobMalling) [inline Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs:300]: Hjælpe metode til Asserts?

File: Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs [hunk 5]
```diff
@@ -235,6 +249,284 @@ public void Cannot_Create_With_Parent_If_Different_Organizations()
...
+            ExpectGetOptionTypeReturnsIfInputIdIsDefined<ItContractTemplateType>(organizationUuid, contractTemplateId, operationError);
+
+            // Act
+            var result = _sut.Create(organizationUuid, itContractModificationParameters);
+
+            // Assert
+            AssertFailureWithKnownErrorDetails(result, $"Failure while resolving ItContractTemplateType option:{operationError.Message.GetValueOrEmptyString()}", operationError.FailureType, transaction);
+        }
+
+        //TODO: from here
```
Comment (JacobMalling) [inline Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs:476]: TODO?

**Review by JacobMalling — APPROVED** [review]
Nogle få rettelser

File: Core.ApplicationServices/Contract/Write/ItContractWriteService.cs [hunk 6]
```diff
@@ -91,10 +103,130 @@ public class ItContractWriteService : IItContractWriteService
...
+                if (resultValue.available == false && contract.AssociatedAgreementElementTypes.Any(x => x.AgreementElementType.Uuid == uuid) == false)
+                {
+                    return new OperationError($"Tried to add agreement element which is not available in the organization: {uuid}", OperationFailure.BadInput);
+                }
+                agreementElementTypes.Add(resultValue.option);
+            }
+
+            var before = contract.AssociatedAgreementElementTypes.ToList();
+
+            var error = contract.SetAgreementElements(agreementElementTypes);
```
[the remark this answers is not in the document]
Comment (mrjsawdk) [reply Core.ApplicationServices/Contract/Write/ItContractWriteService.cs:157]: Det er en domæneregel og den er implementeret i domænemodellen (ItContract)

File: Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs (hunk 2 above)
Comment (mrjsawdk) [reply Tests.Integration.Presentation.Web/Contract/V2/ItContractsApiV2Test.cs:679]: Det gør det muligt at komme med to forskellige lhs og rhs typer så længe de samme type constraints overholdes

File: Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs (hunk 5 above)
Comment (mrjsawdk) [reply Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs:476]: den skal dø

File: Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs (hunk 4 above)
Comment (mrjsawdk) [reply Tests.Unit.Core.ApplicationServices/ApplicationServices/Contract/ItContractWriteServiceTest.cs:300]: tjo det skal vi nok have

**Pull request merged**